coap-core/src/main/java/com/mbed/coap/server/internal/DuplicationDetector.java

```diff
@@ -73,21 +62,27 @@ public DuplicationDetector(TimeUnit unit,
             long cleanIntervalMillis,
             long warningMessageIntervalMillis,
             ScheduledExecutorService scheduledExecutor) {
+        this(unit, duplicationTimeout, new CacheImpl("CoAP duplicate detection cache",
+                maxSize,
+                cleanIntervalMillis,
+                warningMessageIntervalMillis,
+                scheduledExecutor));
+        LOGGER.debug("Duplicate detector: init (max traffic: " + (int) (maxSize / (duplicationTimeout / 1000.0d)) + " msg/sec)");
+    }
+
+    public DuplicationDetector(TimeUnit unit,
```

Comment:
I think this class should be this simple:
```
public class DuplicationDetector {
    public static final CoapPacket EMPTY_COAP_PACKET = new CoapPacket(null);

    private final ConcurrentMap<CoapRequestId, CoapPacket> requestMap;

    public DuplicationDetector(ConcurrentMap<CoapRequestId, CoapPacket> requestMap) {
        this.requestMap = requestMap;
    }

    public CoapPacket isMessageRepeated(CoapPacket request) {
        CoapRequestId requestId = new CoapRequestId(request.getMessageId(), request.getRemoteAddress());

        CoapPacket resp = requestMap.putIfAbsent(requestId, EMPTY_COAP_PACKET);
        return resp;
    }

    void putResponse(CoapPacket request, CoapPacket response) {
        CoapRequestId requestId = new CoapRequestId(request.getMessageId(), request.getRemoteAddress());
        requestMap.put(requestId, response);
    }

    static class CoapRequestId {

        private final int mid;
        private final InetSocketAddress sourceAddress;

        public CoapRequestId(int mid, InetSocketAddress sourceAddress) {
            this.mid = mid;
            this.sourceAddress = sourceAddress;
        }

        @Override
        public boolean equals(Object obj) {
            if (this == obj) {
                return true;
            }
            if (obj == null || getClass() != obj.getClass()) {
                return false;
            }

            CoapRequestId objRequestId = (CoapRequestId) obj;

            if (mid != objRequestId.mid) {
                return false;
            }
            return sourceAddress != null ? sourceAddress.equals(objRequestId.sourceAddress) : objRequestId.sourceAddress == null;
        }

        @Override
        public int hashCode() {
            int result = mid;
            result = 31 * result + (sourceAddress != null ? sourceAddress.hashCode() : 0);
            return result;
        }
    }
}
```

And maybe CoapRequestId should be extracted to it's own file.

Reply:
It would also mean that `setCleanIntervalMillis` is not needed on the cache interface.

Reply:
```
private final ConcurrentMap<CoapRequestId, CoapPacket> requestMap;
```
If the cache is like that, how to implement expiry for the default implementation? Where is that done?

Reply:
You could do that even the same way as before, but as a separate implementation that implements ConcurrentMap instead of this Cache interface you have defined here. So DuplicateDetector should not care about how it's done. So it shouldn't call those methods to handle expiration.